### PR TITLE
TOTEM nT2 unpacker

### PR DESCRIPTION
**Pull request by forthommel**
#### PR description:

We introduce the unpacker and DIGI-level data format for the new TOTEM T2 telescope (or nT2) to operate in Autumn 2022 special pp and HI runs.
nT2 digis are reflecting the registers commonly used for detectors with timing capability: channel, GEOid, and channel status marker for the labelling, and leading+trailing edges measurements for the timing part.
Additionally, a new DetId object (derived from CTPPSDetId to avoid code duplication) is introduced for the mapping of all channels/tiles in each plane of both arms.

This new detector is assigned the FED ID 577, initially used along run 2 as a TOTEM trigger counter and since then unused. Therefore, it replaces this counter unpacking in the PPS standard sequence. For the purpose of the end of 2022 HI run a nT2-only sequence will be introduced in a follow-up PR including the DIGI-to-RecHit conversion (thanks to the addition and usage of the geometry CondFormat linking digis to spatial coordinates).

Finally, work is ongoing on the DQM component, and will also be the object of a follow-up PR.

#### PR validation:

A fake mapping of nT2 channels was introduced to redirect the PPS diamond detector (sharing a common VFAT frame payload) into the nT2 raw-to-digi unpacker. Although the diamond data format does not follow nT2's the bitwise readout of all frames was performed successfully following the data format implemented by TOTEM DAQ experts.
This mapping is left as an example, to be replaced with the actual one after the detector installation, and converted into a payload to be GT'ed.

Furthermore, the usual matrix tests passed flawlessly.

### CODE REVIEW

**Comment by ChrisMisan** [comment]
type ctpps, new-feature

**Comment by ChrisMisan** [comment]
please test

**Comment by ChrisMisan** [comment]
please abort

**Comment by ChrisMisan** [comment]
test parameters:
- workflow 1041

**Comment by ChrisMisan** [comment]
please test

**Comment by ChrisMisan** [comment]
@forthommel this PR will probably get moved to the 12_6_X and then it'll have to be backported to the 12_5_X

**Comment by forthommel** [comment]
@ChrisMisan Let us see if the limited changes for the rest of the CMS environment will make it shorter than the end of this pre-releases cycle!

**Comment by ChrisMisan** [comment]
please abort

- abort requested by the author

**Comment by ChrisMisan** [comment]
please test

File: DataFormats/TotemReco/interface/TotemT2Digi.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ *   Laurent Forthomme
+ *
+ ****************************************************************************/
+
+#ifndef DataFormats_TotemReco_TotemT2Digi_h
+#define DataFormats_TotemReco_TotemT2Digi_h
+
+class TotemT2Digi {
+public:
+  explicit TotemT2Digi() = default;
```
Comment (ChrisMisan) [inline DataFormats/TotemReco/interface/TotemT2Digi.h:14]: Can you explain, why is this constructor explicit?

Comment (forthommel) [reply DataFormats/TotemReco/interface/TotemT2Digi.h:14]: No particular reason, probably a leftover from a previous constructor with a single VFAT frame argument. Can be removed in a later push.

**Comment by forthommel** [comment]
Interesting! Could not reproduce the unit test failure on my workbench.
```
---> test testSiStripDQM_OfflineTkMap succeeded
TestTime:1203
^^^^ End Test testSiStripDQM_OfflineTkMap ^^^^
>> Tests for package DQM/SiStripMonitorClient ran.
```

**Comment by ChrisMisan** [comment]
@forthommel This is probably unrelated to the changes in this pr, #38885

**Comment by ChrisMisan** [comment]
@forthommel can you please squash the commits?

**Comment by forthommel** [comment]
> @forthommel can you please squash the commits?

@ChrisMisan I would rather not, for the same reason as usual: keep track of the development history.

**Comment by ChrisMisan** [comment]
> > @forthommel can you please squash the commits?
> 
> @ChrisMisan I would rather not, for the same reason as usual: keep track of the development history.

sure, but you can rebase it to couple of commits, right? I am talking about commits like "code-format", which don't contain any useful information.

**Comment by forthommel** [comment]
Done. Did not reach the "couple of commits", but already severely broke enough causality for today :)
Depending on the current reviewers' internal policy I believe the best option if you want to reduce drastically the number of commits is to eventually move to a "squash-and-merge" strategy at the final stage.

**Comment by ChrisMisan** [comment]
Thanks @forthommel

**Comment by ChrisMisan** [comment]
please test

**Comment by perrotta** [comment]
> Done. Did not reach the "couple of commits", but already severely broke enough causality for today :) Depending on the current reviewers' internal policy I believe the best option if you want to reduce drastically the number of commits is to eventually move to a "squash-and-merge" strategy at the final stage.

That would be much easier also for us, the release managers. However, I imagine that, besides a few exceptional situations, one wants to let the authors aware of what was done with their PR, and able to check the final results of the operation before merging.

I think that a possible "squash-without-merging" botton could be extremely more useful here. Maybe something we can think about with @cms-sw/core-l2 , or @smuzaffar (as soon as he's back from vacations)

**Comment by smorovic** [comment]
+daq

**Comment by forthommel** [comment]
Spurious unit test failure absent in `CMSSW_12_5_X_2022-07-27-2300` but observed from `CMSSW_12_5_X_2022-07-28-1100` on, compatible with #38885.

**Comment by ChrisMisan** [comment]
+alca

- unpacker tested with 1041
- differences in line with the PR changes
- unit test fail is unrelated to the changes in this PR

**Comment by forthommel** [comment]
It looks like the issue documented in #38885 is fixed now. Maybe (yet) another extra check can be triggered to clear the "tests-rejected" flag?

**Comment by tvami** [comment]
> It looks like the issue documented in https://github.com/cms-sw/cmssw/issues/38885 is fixed now.

Where is the fix? I dont think it got fixed yet...

**Comment by forthommel** [comment]
At least it was not spotted in the last battery of tests launched in #38280 (and a couple of other PRs in review)

**Comment by makortel** [comment]
> At least it was not spotted in the last battery of tests launched in #38280 (and a couple of other PRs in review)

We run only the tests in the packages of the PR and in packages that depend on those.

**Comment by forthommel** [comment]
With the merge of #38905 I believe the IB tests can be triggered once again?

**Comment by ChrisMisan** [comment]
please test

**Comment by tvami** [comment]
@cmsbuild , please test with https://github.com/cms-sw/cmssw/pull/38905

**Comment by civanch** [comment]
+1

**Comment by forthommel** [comment]
Glad to see no rebase is needed to include #38905.
@jpata, @clacaputo, any help needed in your review?

**Comment by smuzaffar** [comment]
> I think that a possible "squash-without-merging" botton could be extremely more useful here. Maybe something we can think about with @cms-sw/core-l2 , or @smuzaffar (as soon as he's back from vacations)

@perrotta , cms bot rely on PR merge commit to properly track the merged PRs.  `squash-without-merging` does not generate the extra merge commit and this can cause bot to not show such PRs in release notes and IB pages

**Comment by forthommel** [comment]
More food for thought that clearly go beyond the scope of this very simple PR :)
@jpata, @clacaputo, any help needed in your review?

**Comment by clacaputo** [comment]
+reconstruction

**Comment by qliphy** [comment]
please test
to refresh

**Comment by forthommel** [comment]
It seems that everything remains stable wrt the previous battery of tests :)

**Comment by qliphy** [comment]
+1

**Pull request merged**